### PR TITLE
taproot: Split errors up

### PR DESCRIPTION
**Pull request by tcharding**
Currently there are a couple of errors in the `taproot` module that are too general, resulting in functions that return a general error type when a specific one would do.
    
Split two errors out and use them for for enum variants and function returns as possible.
 
Done as part of #2883

### CODE REVIEW

**Comment by apoelstra** [comment]
In 1be14146dcef2c306b5d6b5d7390fb80047088dd

Both `Taproot` and `Merkle` are proper nouns IMO and should stay capitalized.

**Comment by apoelstra** [comment]
989595b270aa4808905fbf66ca2812bbf842ed51 looks good otherwise! Definitely an improvement.

**Comment by tcharding** [comment]
Hey @jamillambert this is a pretty standard error cleanup PR, if you are interested in the direction we are going with regards to error handling this PR would be a good one to review and understand. If you find error handling boring then don't sweat it, its mainly interesting as a pure engineering problem because there is a whole bunch of nuance in how the Rust language treats enums, enum variant fields, and `From` impls in relation to public API surface which effects semver. For 1.0 we want to be confident that we can update code without accidentally breaking semver. If all that is over your head don't sweat it, just mentioning it in case you find it interesting.

**Comment by tcharding** [comment]
> Both `Taproot` and `Merkle` are proper nouns IMO and should stay capitalized.

I was/am unsure whether it should be "Merkle tree" or "Merkle Tree" - I went with the later. Fixed also "Huffman Tree".

Please note there is one error message changed in the first patch to remove the two sentences.

**Comment by storopoli** [comment]
> I was/am unsure whether it should be "Merkle tree" or "Merkle Tree" - I went with the later. Fixed also "Huffman Tree".

In the docstring can be "Merkle tree" (see [Wikipedia - Merkle tree](https://en.wikipedia.org/wiki/Merkle_tree) for example).
In the enum variants it should be `MerkleTree`.
So, in my opinion, your approach is correct.

**Review by jamillambert — COMMENTED** [review]
I am not sure I understand completely why only 2 of the 5 `TaprootError` variants were split out.  Is it because of the functions that only return one of these,  whereas the other variants are returned by functions that can return other variants too and therefore need to be general?

But in terms of achieving what is described I don't see any issues.

File: bitcoin/src/taproot/mod.rs [hunk 1]
```diff
@@ -667,7 +667,7 @@ impl core::fmt::Display for HiddenNodesError {
 
         f.write_str(match self {
             HiddenParts(_) =>
-                "an attempt to construct a tap tree from a node_info containing hidden parts.",
+                "an attempt to construct a tap tree from a node_info containing hidden parts",
```
[the remark this answers is not in the document]
Comment (apoelstra) [reply bitcoin/src/taproot/mod.rs:670]: Either that or "taptree" which is one word.

File: bitcoin/src/taproot/mod.rs [hunk 2]
```diff
@@ -921,26 +921,26 @@ impl<'de> serde::Deserialize<'de> for NodeInfo {
...
+                                    "hidden node insertion error: {}",
                                     e
                                 ))
                             })?;
                         }
                     }
                 }
                 NodeInfo::try_from(builder).map_err(|e| {
-                    serde::de::Error::custom(format!("Incomplete taproot tree: {}", e))
+                    serde::de::Error::custom(format!("incomplete taproot tree: {}", e))
```
[the remark this answers is not in the document]
Comment (tcharding) [reply bitcoin/src/taproot/mod.rs:943]: Legend, thanks. Will fix.

**Comment by tcharding** [comment]
> I am not sure I understand completely why only 2 of the 5 `TaprootError` variants were split out. Is it because of the functions that only return one of these, whereas the other variants are returned by functions that can return other variants too and therefore need to be general?

Yes that is correct. The idea, and if you notice this violated anywhere it should be fixed, is that for every function that returns an enum error type all variants of the error type are returned by the function. This is something that we have been working towards over the last year or so, it is highly likely there are still some that need fixing.

**Comment by tcharding** [comment]
I've dropped the first patch that was attempting to fix the capitalization of error messages because there are a whole bunch of anomalies in the `taproot` module around spelling and capitalization of Taproot tree.

**Comment by tcharding** [comment]
Added https://github.com/rust-bitcoin/rust-bitcoin/issues/2913 to track the spelling/capitalization stuff that was dropped.

File: bitcoin/src/taproot/mod.rs [hunk 3]
```diff
@@ -1348,13 +1348,7 @@ impl fmt::Display for TaprootBuilderError {
...
-            InvalidMerkleTreeDepth(d) => {
-                write!(
-                    f,
-                    "Merkle Tree depth({}) must be less than {}",
-                    d, TAPROOT_CONTROL_MAX_NODE_COUNT
-                )
-            }
+            InvalidMerkleTreeDepth(ref e) => write_err!(f, "invalid Merkle tree depth"; e),
             NodeNotInDfsOrder => {
                 write!(f, "add_leaf/add_hidden must be called in DFS walk order",)
```
Comment (Kixunil) [inline bitcoin/src/taproot/mod.rs:1353]: Unrelated but I think error messages should say what's wrong not how it should be right.

File: bitcoin/src/taproot/mod.rs [hunk 4]
```diff
@@ -1375,22 +1369,27 @@ impl std::error::Error for TaprootBuilderError {
...
     /// Proof size must be a multiple of 32.
     InvalidMerkleBranchSize(usize),
     /// Merkle tree depth must not be more than 128.
-    InvalidMerkleTreeDepth(usize),
+    InvalidMerkleTreeDepth(InvalidMerkleTreeDepthError),
     /// The last bit of tapleaf version must be zero.
-    InvalidTaprootLeafVersion(u8),
+    InvalidTaprootLeafVersion(InvalidTaprootLeafVersionError),
     /// Invalid control block size.
     InvalidControlBlockSize(usize),
```
Comment (Kixunil) [inline bitcoin/src/taproot/mod.rs:1394]: Might be nice to newtype the other variants too.

File: bitcoin/src/taproot/mod.rs [hunk 5]
```diff
@@ -1439,15 +1432,65 @@ impl std::error::Error for TaprootError {
...
 
+impl From<InvalidTaprootLeafVersionError> for TaprootError {
+    fn from(e: InvalidTaprootLeafVersionError) -> Self { Self::InvalidTaprootLeafVersion(e) }
+}
+
+impl From<InvalidMerkleTreeDepthError> for TaprootError {
+    fn from(e: InvalidMerkleTreeDepthError) -> Self { Self::InvalidMerkleTreeDepth(e) }
+}
+
+/// Merkle tree depth must not be more than 128.
```
Comment (Kixunil) [inline bitcoin/src/taproot/mod.rs:1450]: This doc comment is unrelated.

**Review by Kixunil — COMMENTED** [review]
Other than the doc comment issue ACK 12baa0b891. Feel free to one-ACK after it's fixed without other changes.

File: bitcoin/src/taproot/mod.rs (hunk 3 above)
Comment (tcharding) [reply bitcoin/src/taproot/mod.rs:1353]: Ah, interesting. I think all the error messages in general need work but definitely the ones in `taproot`.

File: bitcoin/src/taproot/mod.rs (hunk 4 above)
Comment (tcharding) [reply bitcoin/src/taproot/mod.rs]: Added an additional patch, good reviewing - thanks.

File: bitcoin/src/taproot/mod.rs (hunk 5 above)
Comment (tcharding) [reply bitcoin/src/taproot/mod.rs]: Woops, fixed.

File: bitcoin/src/taproot/mod.rs (hunk 3 above)
Comment (tcharding) [reply bitcoin/src/taproot/mod.rs:1353]: Left for another day.

**Comment by tcharding** [comment]
Needs ack from @Kixunil please because changes since last ack are non-trivial. No rush on this one its just error handling stuff.

**Review by Kixunil — DISMISSED** [review]
ACK 0c9223ac0508eceea40aeb09dc6d1fc36669054b

**Review by apoelstra — APPROVED** [review]
ACK 6a7f78001845b70fe415117ff7dbca4d5c97c01d
